src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs

```diff
@@ -456,14 +466,69 @@ private static HeadersSyncBatch BuildDependentBatch(HeadersSyncBatch batch, long
             dependentBatch.StartNumber = addedEarliest;
             int count = (int)(addedLast - addedEarliest + 1);
             dependentBatch.RequestSize = count;
-            dependentBatch.MinNumber = batch.MinNumber;
             dependentBatch.Response = batch.Response!
                 .Skip((int)(addedEarliest - batch.StartNumber))
                 .Take(count).ToPooledList(count);
             dependentBatch.ResponseSourcePeer = batch.ResponseSourcePeer;
             return dependentBatch;
         }
 
+        private void EnqueueBatch(HeadersSyncBatch batch)
+        {
+            HeadersSyncBatch? left = ProcessPersistedPortion(batch);
+            if (left is not null)
+            {
+                _pending.Enqueue(batch);
+            }
+        }
+
+        /// <summary>
+        /// Check for portion of header that is already persisted and process them, returning a null batch
+        /// if the whole portion is already persisted and does not require download.
+        /// If only portion of the batch is persisted, then return a new batch that need to be downloaded.
+        /// </summary>
+        /// <param name="batch"></param>
+        /// <returns></returns>
+        private HeadersSyncBatch? ProcessPersistedPortion(HeadersSyncBatch batch)
+        {
+            // This only check for the last header though, which is fine as headers are so small, the time it take
+            // to download one is more or less the same as the whole batch. So many small batch is slower than
+            // less large batch.
+            BlockHeader? lastHeader = _blockTree.FindHeader(batch.EndNumber, BlockTreeLookupOptions.None);
+            if (lastHeader is null) return batch;
+
+            using ArrayPoolList<BlockHeader> headers = new ArrayPoolList<BlockHeader>(1);
+            headers.Add(lastHeader);
+            for (long i = batch.EndNumber - 1; i >= batch.StartNumber; i--)
+            {
+                // Don't worry about fork, `InsertHeaders` will check for fork and retry if it is not on the right fork.
+                BlockHeader nextHeader = _blockTree.FindHeader(lastHeader.ParentHash!, i);
+                if (nextHeader is null) break;
+                lastHeader = nextHeader;
+                headers.Add(lastHeader);
+            }
+
+            ArrayPoolList<BlockHeader> reversedBatch = new ArrayPoolList<BlockHeader>(headers.Count);
+            foreach (BlockHeader blockHeader in headers.Reverse())
+            {
+                reversedBatch.Add(blockHeader);
+            }
```

Comment:
Why need a second `ArrayPoolList`? Add `Reverse` to the class and reverse in-place (same as `List<>`)

Reply:
Changed to `AsSpan().Reverse()`.